### DISCUSSION TITLE
Anything like trace.Span.SetAttributes() for metrics?

### DISCUSSION THREAD

**Discussion by JobberRT** [open-telemetry/opentelemetry-go#4208] [Q&A] [answered]
## Usage example:

There are some mission with different name, and their job is to do http request. I  use `Int64Gauge`(Because `Prometheus` Gauge is not overlap and cumulative) to record every HTTP STATUS CODE's count for each mission. In the `Prometheus` SDK, I can pass a slice of string representing the labels to achieve this, but in OTel, I have no way but formating it as a string which is complex for resolving.

## What I have got:
- I use trace too, and I found `trace.Span.SetAttributes` is maybe what I need.
- I found from both [Official Doc](https://opentelemetry.io/docs/specs/otel/overview/#metric-signal) and SDK code that I can only set attributes for `Meter`, and all `Instrument` inherit it

## Another thing....
Is there any suggest stack of OTel deployment? For now I'm using `OTel-Collector`+`Grafana`+`Jaeger`+`Prometheus`, but `Grafana` isn't design for `OTel` and I have to use `OTel-Collector` to transform data to specify format and send them to different backend then use `Grafana` pull from each backend.

I tried `SigNoz` but its ui is not powerful enough.

I would like these following features:
- Alert isn't necessary because I have alert based on ELK
- UI like `Grafana`(powerful enough) and work like `SigNoz`(native support for otel and has storage backend included, signoz use clickhouse for example)

**Comment by dmathieu** [comment]
You can set specific attributes to a metric when emitting it.

For example:

```golang
counter, _ := meter.Int64Counter("test")
counter.Add(context.Background(), 1, metric.WithAttributes(
	attribute.String("foo", "bar"),
))
```

However, you can't add attributes to a metric once it has been emitted

**Reply by JobberRT** [reply]
Thanks dmathieu, I tried `Counter` and `UpDownCounter` they have .Add() method, but it seems that `Gauge` and `Histogram` doesn't have this method. And what I need is period, no overlapping(which means reset every export) with prometheus.

Prometheus's Counter can only go up, I will try something else which supported by Grafana.

**Reply by dmathieu** [reply]
Histograms have a `Record` method, which also takes the same options for arguments.
See https://github.com/open-telemetry/opentelemetry-go/blob/135c64f2897000780febb0945d1d9db195faecc6/metric/syncint64.go#L144

**Reply by JobberRT** [reply]
Ohhh, sorry for the mistake, I was testing all instruments and ended up confusing myself. But `Gauge` indeed doesn't have this kind of method, right?

**Reply by pellared** [reply, marked as answer]
For `Gauge` please refer to https://pkg.go.dev/go.opentelemetry.io/otel/metric#example-Meter-Asynchronous_single and https://pkg.go.dev/go.opentelemetry.io/otel/metric#WithAttributes

**Reply by JobberRT** [reply]
ohhh~ For `Obsverble Instrument`, it is the `Observer` that actually do the "Record" thing, so I can't set attributes to the `Gauge`, instead I should set it to the `Observer`! That make sense!
